### PR TITLE
[clean-job-do-not-use] support get distinct job lasted and clean jobs

### PR DESCRIPTION
**Pull request by truongnguyen101195**

### CODE REVIEW

File: .github/workflows/golangci-lint.yml [hunk 1]
```diff
@@ -4,7 +4,6 @@ on:
     tags:
       - v*
     branches:
-      - master
```
Comment (huylenq) [inline .github/workflows/golangci-lint.yml:7]: Shouldn't if there's no particular reason.

File: job/cache.go [hunk 2]
```diff
@@ -203,27 +203,48 @@ func (c *MemoryJobCache) PersistEvery(persistWaitTime time.Duration) {
...
+		cache.dbCleaning = cleaning
 	}
 }
 
 func (c *LockFreeJobCache) Start(persistWaitTime time.Duration, jobstatTtl time.Duration) {
 	if persistWaitTime == 0 {
 		c.PersistOnWrite = true
 	}
+	//Start Cron job Clean job is done
+	go c.cronClearJobDoNotUse()
```
Comment (khanhngobackend) [inline job/cache.go]: Why put this here instead of below the line `247` below to completely eliminate the possibility of race?

**Review by khanhngobackend — CHANGES_REQUESTED** [review]
Almost all LGTM, except 1 (commented).

File: job/storage/postgres/postgres.go [hunk 3]
```diff
@@ -35,7 +35,12 @@ func New(dsn string) *DB {
 
 // GetAll returns all persisted Jobs.
 func (d DB) GetAll() ([]*job.Job, error) {
-	query := fmt.Sprintf(`select coalesce(json_agg(j.job), '[]'::json) from (select * from %[1]s) as j;`, TableName)
+	query := fmt.Sprintf(`with  clean as (
+select ctid ,
+(job -> 'metadata' -> 'last_attempted_run') as last_run, row_number() over(partition by (job->'id') order by  (job -> 'metadata' -> 'last_attempted_run') desc) as num 
+from %s where CAST(job ->> 'is_done' as boolean) = false
+) 
+select  coalesce(json_agg(j.job), '[]'::json) from %s j where ctid in (select ctid from clean where num = 1);`, TableName, TableName)
```
Comment (huylenq) [inline job/storage/postgres/postgres.go]: A quite surprising behavior, and may yield subtle bugs: "GetAll" does not actually returning "all", but only a subset from this filter. How about return deduplicated result but regardless of the job's completion status to reserve the current semantic?
```suggestion
	query := fmt.Sprintf(
		`with incomplete_jobs as (
			select ctid,
					row_number() over(partition by (job->'id') order by (job->'metadata'->'last_attempted_run') desc) as row_num 
				from %s) 
		select coalesce(json_agg(j.job), '[]'::json)
		from %s j
		where ctid in (select ctid from incomplete_jobs where row_num = 1);`,
		TableName, TableName)
```

**Comment by huylenq** [comment]
Discussed offline with @truongnguyen101195 , @khanhngobackend , @vyduong1207. @truongnguyen101195 will take a more thorough look at how these work.

**Review by huylenq — APPROVED** [review]
The concern of accidentally removing/filtering the latest version of the job record (instead of the stale version) is still there.
However, the jobs data in this service only serve execution purposes, not be referenced for bookkeeping/management sake, so losing records is actually not a big deal. So we can merge this, but please be aware for any issue related to job history, if they ever surface.
@truongnguyen101195, @vyduong120, @khanhngobackend

**Pull request merged**